### ISSUE TITLE
probe: separation

### ISSUE THREAD

**Issue by leondz** [NVIDIA/garak#579]
"Interesting way of looking at the difference in conditional distribution when it interprets as instruction vs as data. Might be something there to tease out and fit into garak somehow maybe."

see https://arxiv.org/abs/2403.06833

**Comment by leondz** [comment]
instruction hierarchy strength test, still seems worth investigating

**Comment by adelzaitri** [comment]
@leondz 
I'd like to pick this one up.
Or is anyone already working on this?

If it's free, my next step is to post a proposed way to address it (the paper's measure is a *difference between two conditions* rather than a per-attempt hit rate (garak's), so the interesting question is how it should sit inside garak's probe → detector → pass rate model.

**Comment by adelzaitri** [comment]
@jmartin-tech @patriciapampanelli 
Do you know if this probe is being worked on somewhere else in the repo? It doesn't look like it from my search.
Happy to start on it and propose some implementation ideas, or pivot to another issue of priority first

**Comment by adelzaitri** [comment]
@leondz 
My initial thoughts on the fit to garak
the fit is good and most of it already exists in garak. SEP's oracle is a witness word per record, so `TriggerListDetector` (`garak/detectors/base.py:277`) scores it as-is with no new detector.

One design question before I write anything. Separation is a difference between two conditions (same probe instruction in the system prompt vs in the data), not a per-attempt hit rate, so it doesn't drop straight into probe → detector → pass rate. Three shapes:

**(a)** Two probes — data-channel scored, instruction-channel INFORMATIONAL. Works today, but garak never computes the difference and the control has inverted semantics.

**(b)** One attempt holding both conditions as two conversations. **Doesn't work** — `attempt.conversations` is N samples of one prompt, not N conditions (`garak/probes/base.py:313`, `garak/attempt.py:426`), and `generations` defaults to 5. Would need an Attempt-model change, so I've ruled it out.

**(c)** One probe, two attempts per record, paired by `attempt.notes["sep_record_id"]`. Each is a normal single-prompt attempt with its own system turn — same construction as `garak/probes/sysprompt_extraction.py:133-139`. Detector unchanged, delta computed at reporting time. No architecture touched.

I'd go with (c). Is that the shape you want, or would you rather I ship (a)?

(Separate, smaller: the 12MB dataset needs either a stratified subsample under `garak/data/` or a `garak-llm` mirror — happy to follow whichever you prefer.)

**Comment by leondz** [comment]
Would this make sense as a system prompt vs standard turn, to match the control/data plane separation framing in the paper?

**Comment by adelzaitri** [comment]
@leondz 
Yes — actual `system` role rather than a formatted prefix. The paper's eval code does the same, and its inference loop calls the model twice per record and pairs afterwards, which is the shape of (c):

https://github.com/egozverev/Should-It-Be-Executed-Or-Processed/blob/main/model_eval/get_model_outputs.py#L104-L121

One catch worth knowing before I build it that way. In `huggingface.py`, when the tokenizer has no chat template, garak sends `prompt.last_message().text` only — `Pipeline:141`, `Model:463`. The system turn is dropped silently. For this probe that doesn't degrade the result, it inverts it: the probe never reaches the model, instruction-channel execution reads zero, and zero-vs-hits scores as *perfect separation*. The SEP authors hit the same wall and hard-coded `gemma2b` / `gemma7b` / `starling` to a concatenated fallback (`get_model_outputs.py:111-117`) — theirs is visible in the results, ours wouldn't be.

Cheap fix stays inside the probe: a control condition per target, probe alone in `system` with trivial data in `user`. No witness there means the system channel isn't reaching that model, so it gets reported rather than scored.

Include that in v1, or leave it out and document the requirement? I'd include it — the failure mode looks like a good score rather than an error.

**Comment by jmartin-tech** [comment]
Commenting on the _catch_ identified:

@adelzaitri, you are correct on your assessment of the `huggingface` generator, in fact many generators lack support for `system` vs `user` content and only send the `user` message. Your solution in this case of using probe level detection seems sound, an initial suitability check against the generator could allow the probe return without executing any prompts and provide signal that the probe is not suitable for use against the target as configured.

As to the other points raised:
* I like option (c) though I suspect it will need some tweaking and it may introduce some consistency concerns in the run's report.jsonl output. During the review process we can likely figure out a path forward. There is currently an expectation in `base.Probe` that any status 1 `Attempt` has a corresponding status 2 `Attempt` which is expanded with the scored result. It may be viable to use `base.IterativeProbe` as the base class since expectations for those probes are evolving and the scoring expectations for them are more varied. I could see a probe for this technique using `base.IterativeProbe` and the `_generate_next_attempts()` method in that probe to get the linked `Attempt` values in queue.
* The dataset likely would be best left as retrieved content, a huggingface mirror with proper attribution may be the simplest path as the authors list them as licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/).

**Comment by adelzaitri** [comment]
Thanks @jmartin-tech for the clarification, that's enough for me to start drafting. I'll go with `IterativeProbe` and `_generate_next_attempts()` for the pair, and add the generator suitability check so an unsuitable target returns before any prompts are posed.

One more thing about the dataset on HF: I assume `garak-llm` will be the right home for the mirror...would you want to upload it, or should I host it under my own namespace and you re-point it later? Either works, not a blocker for now

**Comment by jmartin-tech** [comment]
@adelzaitri go ahead and host it in your namespace and we can fork/clone and point to the project maintained namespace during the review and testing phase.  This enable you to evolve and test the dataset format as you build your draft.
